File: Easy/144. Binary Tree Preorder Traversal/BinaryTreePreorderTraversal.py
'''
LeetCode Easy: 144. Binary Tree Preorder Traversal
https://leetcode.com/problems/binary-tree-preorder-traversal/

Check BinaryTreePreorderTraversal.cpp for brief solution explanation.
'''
class Solution:
    def preorderTraversal(self, root: Optional[TreeNode]) -> List[int]:
        preorder = []
        self.dfs(root, preorder)
        return preorder
    def dfs(self, root, preorder):
        if root==None:
            return
        preorder.append(root.val)
        self.dfs(root.left, preorder)
        self.dfs(root.right, preorder)

File: Easy/144. Binary Tree Preorder Traversal/BinaryTreePreorderTraversal.java
'''
LeetCode Easy: 144. Binary Tree Preorder Traversal
https://leetcode.com/problems/binary-tree-preorder-traversal/

Check BinaryTreePreorderTraversal.cpp for brief solution explanation.
'''
class Solution {
    public List<Integer> preorderTraversal(TreeNode root) {
        List<Integer>preorder = new LinkedList<Integer>();
        dfs(root, preorder);
        return preorder;
    }
    public void dfs(TreeNode root, List<Integer>preorder){
        if(root==null){
            return;
        }
        preorder.add(root.val);
        dfs(root.left, preorder);
        dfs(root.right, preorder);
    }
}